enzyme/Enzyme/TraceInterface.cpp

```diff
@@ -164,105 +222,226 @@ DynamicTraceInterface::DynamicTraceInterface(Value *dynamicInterface,
   }
 
   assert(sampleFunction);
+  assert(dynamicInterface);
+
+  auto &M = *F->getParent();
+  IRBuilder<> Builder(F->getEntryBlock().getFirstNonPHIOrDbg());
+
+  getTraceFunction = MaterializeGetTrace(Builder, dynamicInterface, M);
+  getChoiceFunction = MaterializeGetChoice(Builder, dynamicInterface, M);
+  getLikelihoodFunction =
+      MaterializeGetLikelihood(Builder, dynamicInterface, M);
+  insertCallFunction = MaterializeInsertCall(Builder, dynamicInterface, M);
+  insertChoiceFunction = MaterializeInsertChoice(Builder, dynamicInterface, M);
+  newTraceFunction = MaterializeNewTrace(Builder, dynamicInterface, M);
+  freeTraceFunction = MaterializeFreeTrace(Builder, dynamicInterface, M);
+  hasCallFunction = MaterializeHasCall(Builder, dynamicInterface, M);
+  hasChoiceFunction = MaterializeHasChoice(Builder, dynamicInterface, M);
+
+  assert(getTraceFunction);
+  assert(getChoiceFunction);
+  assert(getLikelihoodFunction);
+  assert(insertCallFunction);
+  assert(insertChoiceFunction);
+  assert(newTraceFunction);
+  assert(freeTraceFunction);
+  assert(hasCallFunction);
+  assert(hasChoiceFunction);
 }
 
-// implemented by enzyme
-Function *DynamicTraceInterface::getSampleFunction() { return sampleFunction; }
-
-// user implemented
-Value *DynamicTraceInterface::getTrace() {
-  if (getTraceFunction)
-    return getTraceFunction;
-
-  IRBuilder<> Builder(F->getEntryBlock().getFirstNonPHIOrDbgOrLifetime());
-
+llvm::GlobalVariable *
+DynamicTraceInterface::MaterializeGetTrace(IRBuilder<> &Builder,
+                                           Value *dynamicInterface, Module &M) {
   auto ptr = Builder.CreateInBoundsGEP(Builder.getInt8PtrTy(), dynamicInterface,
                                        Builder.getInt32(0));
+  auto pty = PointerType::getUnqual(getTraceTy());
```

Comment:
We should update these pointer types to not be unqual, but based off of the addrspace of origin, if available.

Reply:
would one of these address spaces do the trick?
```
  M.getDataLayout().getAllocaAddrSpace();
  M.getDataLayout().getProgramAddressSpace();
  M.getDataLayout().getNonIntegralAddressSpaces();
  M.getDataLayout().getDefaultGlobalsAddressSpace();
```

Reply:
Since pty is being used on the cast, can you do PointerType::get(getTraceTy(), cast<PointerType>(load->getType())->getAddressSpace()) ?

Reply:
done